File: sdk-ts/src/providers/deepgram-stt.ts
import WebSocket from 'ws';
import { getLogger } from '../logger';

export interface Transcript {
  readonly text: string;
  readonly isFinal: boolean;
  readonly confidence: number;
}

type TranscriptCallback = (transcript: Transcript) => void;

const DEEPGRAM_WS_URL = 'wss://api.deepgram.com/v1/listen';

/**
 * Optional tuning knobs for Deepgram live transcription.
 *
 * Mirrors Python's ``DeepgramSTT`` kwargs so callers can lower turn latency
 * without monkey-patching (BUG #13).
 */
export interface DeepgramSTTOptions {
  /** Model name. Default ``nova-3``. */
  readonly model?: string;
  /** Audio encoding (``linear16`` | ``mulaw`` | etc). Default ``linear16``. */
  readonly encoding?: string;
  /** Sample rate in Hz. Default ``16000``. */
  readonly sampleRate?: number;
  /**
   * Voice-activity endpointing threshold in milliseconds.
   * Lower values reduce turn latency at the cost of more false-start cuts.
   * Default ``150``.
   */
  readonly endpointingMs?: number;
  /**
   * End-of-utterance silence window in milliseconds. Deepgram enforces a
   * hard minimum of 1000 ms. Set to ``null`` to disable. Default ``1000``.
   */
  readonly utteranceEndMs?: number | null;
  /** Enable smart formatting (punctuation + numerals). Default ``true``. */
  readonly smartFormat?: boolean;
  /** Emit interim (non-final) transcripts. Default ``true``. */
  readonly interimResults?: boolean;
  /** Emit VAD events (``SpeechStarted`` / ``UtteranceEnd``). Default ``true``. */
  readonly vadEvents?: boolean;
}

export class DeepgramSTT {
  private ws: WebSocket | null = null;
  private callbacks: TranscriptCallback[] = [];
  /** Request ID from Deepgram — used to query actual cost post-call. */
  requestId: string = '';

  private readonly apiKey: string;
  private readonly language: string;
  private readonly model: string;
  private readonly encoding: string;
  private readonly sampleRate: number;
  private readonly endpointingMs: number;
  private readonly utteranceEndMs: number | null;
  private readonly smartFormat: boolean;
  private readonly interimResults: boolean;
  private readonly vadEvents: boolean;

  /**
   * New ergonomic constructor accepting an options object (mirrors Python kwargs).
   *
   * Also accepts the legacy positional form
   * ``(apiKey, language?, model?, encoding?, sampleRate?)`` for backward
   * compatibility with code that predated BUG #13.
   */
  constructor(
    apiKey: string,
    language?: string,
    model?: string,
    encoding?: string,
    sampleRate?: number,
    options?: DeepgramSTTOptions,
  );
  constructor(apiKey: string, options: DeepgramSTTOptions & { language?: string });
  constructor(
    apiKey: string,
    languageOrOptions?: string | (DeepgramSTTOptions & { language?: string }),
    model?: string,
    encoding?: string,
    sampleRate?: number,
    options?: DeepgramSTTOptions,
  ) {
    this.apiKey = apiKey;
    const opts: DeepgramSTTOptions & { language?: string } =
      typeof languageOrOptions === 'object' && languageOrOptions !== null
        ? languageOrOptions
        : options ?? {};

    this.language = (typeof languageOrOptions === 'string' ? languageOrOptions : opts.language) ?? 'en';
    this.model = model ?? opts.model ?? 'nova-3';
    this.encoding = encoding ?? opts.encoding ?? 'linear16';
    this.sampleRate = sampleRate ?? opts.sampleRate ?? 16000;
    this.endpointingMs = opts.endpointingMs ?? 150;
    this.utteranceEndMs = opts.utteranceEndMs === null ? null : opts.utteranceEndMs ?? 1000;
    this.smartFormat = opts.smartFormat ?? true;
    this.interimResults = opts.interimResults ?? true;
    this.vadEvents = opts.vadEvents ?? true;
  }

  /** Factory for Twilio calls — mulaw 8 kHz. Forwards tuning options through. */
  static forTwilio(
    apiKey: string,
    language: string = 'en',
    model: string = 'nova-3',
    options: DeepgramSTTOptions = {},
  ): DeepgramSTT {
    return new DeepgramSTT(apiKey, language, model, 'mulaw', 8000, options);
  }

  async connect(): Promise<void> {
    const params = new URLSearchParams({
      model: this.model,
      language: this.language,
      encoding: this.encoding,
      sample_rate: String(this.sampleRate),
      channels: '1',
      interim_results: this.interimResults ? 'true' : 'false',
      endpointing: String(this.endpointingMs),
      smart_format: this.smartFormat ? 'true' : 'false',
      vad_events: this.vadEvents ? 'true' : 'false',
      no_delay: 'true',
    });
    if (this.utteranceEndMs !== null) {
      // Deepgram enforces a hard minimum of 1000 ms on this knob.
      params.set('utterance_end_ms', String(Math.max(this.utteranceEndMs, 1000)));
    }

    const url = `${DEEPGRAM_WS_URL}?${params.toString()}`;

    this.ws = new WebSocket(url, {
      headers: { Authorization: `Token ${this.apiKey}` },
    });

    await new Promise<void>((resolve, reject) => {
      const timer = setTimeout(() => reject(new Error('Deepgram connect timeout')), 10000);
      this.ws!.once('open', () => {
        clearTimeout(timer);
        resolve();
      });
      this.ws!.once('error', (err) => {
        clearTimeout(timer);
        reject(err);
      });
    });

    this.ws.on('message', (raw) => {
      let data: {
        type: string;
        is_final?: boolean;
        speech_final?: boolean;
        channel?: { alternatives?: Array<{ transcript?: string; confidence?: number }> };
      };
      try {
        data = JSON.parse(raw.toString()) as typeof data;
      } catch {
        return;
      }

      if (data.type === 'Metadata' && (data as Record<string, unknown>).request_id) {
        this.requestId = (data as Record<string, unknown>).request_id as string;
        return;
      }

      if (data.type !== 'Results') return;

      const alternatives = data.channel?.alternatives ?? [];
      if (!alternatives.length) return;

      const best = alternatives[0];
      const text = (best.transcript ?? '').trim();
      if (!text) return;

      // BUG #13 — ``is_final`` alone marks a stable utterance;
      // ``speech_final`` is a faster end-of-utterance hint from Deepgram's
      // VAD. Accept either so the pipeline doesn't wait up to
      // utterance_end_ms on every turn.
      const transcript: Transcript = {
        text,
        isFinal: Boolean(data.is_final) || Boolean(data.speech_final),
        confidence: best.confidence ?? 0,
      };

      for (const cb of this.callbacks) {
        cb(transcript);
      }
    });
  }

  sendAudio(audio: Buffer): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    this.ws.send(audio);
  }

  onTranscript(callback: TranscriptCallback): void {
    if (this.callbacks.length >= 10) {
      getLogger().warn('DeepgramSTT: maximum of 10 onTranscript callbacks reached; replacing the last callback.');
      this.callbacks[this.callbacks.length - 1] = callback;
      return;
    }
    this.callbacks.push(callback);
  }

  close(): void {
    if (this.ws) {
      try {
        this.ws.send(JSON.stringify({ type: 'CloseStream' }));
      } catch {
        // ignore
      }
      this.ws.close();
      this.ws = null;
    }
  }
}

File: sdk-py/getpatter/providers/deepgram_stt.py
import json
from typing import AsyncIterator
from urllib.parse import urlencode

import websockets

from getpatter.providers.base import STTProvider, Transcript

DEEPGRAM_WS_URL = "wss://api.deepgram.com/v1/listen"


class DeepgramSTT(STTProvider):
    def __init__(
        self,
        api_key: str,
        language: str = "en",
        model: str = "nova-3",
        encoding: str = "linear16",
        sample_rate: int = 16000,
        *,
        endpointing_ms: int = 150,
        utterance_end_ms: int | None = 1000,
        smart_format: bool = True,
        interim_results: bool = True,
        vad_events: bool = True,
    ):
        self.api_key = api_key
        self.language = language
        self.model = model
        self.encoding = encoding
        self.sample_rate = sample_rate
        self.endpointing_ms = endpointing_ms
        self.utterance_end_ms = utterance_end_ms
        self.smart_format = smart_format
        self.interim_results = interim_results
        self.vad_events = vad_events
        self._ws = None
        self.request_id: str | None = None

    def __repr__(self) -> str:
        return f"DeepgramSTT(model={self.model!r}, language={self.language!r}, encoding={self.encoding!r})"

    @classmethod
    def for_twilio(
        cls,
        api_key: str,
        language: str = "en",
        model: str = "nova-3",
        **kwargs,
    ):
        """Create a Deepgram adapter configured for Twilio mulaw 8kHz."""
        return cls(
            api_key=api_key,
            language=language,
            model=model,
            encoding="mulaw",
            sample_rate=8000,
            **kwargs,
        )

    async def connect(self) -> None:
        params = {
            "model": self.model,
            "language": self.language,
            "encoding": self.encoding,
            "sample_rate": str(self.sample_rate),
            "channels": "1",
            "interim_results": "true" if self.interim_results else "false",
            "endpointing": str(self.endpointing_ms),
            "smart_format": "true" if self.smart_format else "false",
            "vad_events": "true" if self.vad_events else "false",
            "no_delay": "true",
        }
        if self.utterance_end_ms is not None:
            # utterance_end_ms has a hard minimum of 1000 on Deepgram's API.
            params["utterance_end_ms"] = str(max(int(self.utterance_end_ms), 1000))
        url = f"{DEEPGRAM_WS_URL}?{urlencode(params)}"
        self._ws = await websockets.connect(
            url,
            additional_headers={"Authorization": f"Token {self.api_key}"},
        )

    async def send_audio(self, audio_chunk: bytes) -> None:
        if self._ws is None:
            raise RuntimeError("Not connected. Call connect() first.")
        await self._ws.send(audio_chunk)

    def _parse_message(self, raw_message: str) -> Transcript | None:
        data = json.loads(raw_message)
        msg_type = data.get("type", "")

        if msg_type == "Metadata":
            self.request_id = data.get("request_id")
            return None

        if msg_type != "Results":
            return None

        alternatives = data.get("channel", {}).get("alternatives", [])
        if not alternatives:
            return None

        best = alternatives[0]
        text = best.get("transcript", "").strip()
        if not text:
            return None

        # is_final alone marks a stable utterance; speech_final is a faster
        # end-of-utterance hint from Deepgram's VAD. Accept either so the
        # pipeline doesn't wait up to utterance_end_ms on every turn.
        is_final = bool(data.get("is_final", False) or data.get("speech_final", False))
        return Transcript(
            text=text,
            is_final=is_final,
            confidence=best.get("confidence", 0.0),
        )

    async def receive_transcripts(self) -> AsyncIterator[Transcript]:
        if self._ws is None:
            raise RuntimeError("Not connected. Call connect() first.")

        async for raw_message in self._ws:
            if isinstance(raw_message, bytes):
                continue  # Skip binary frames
            transcript = self._parse_message(raw_message)
            if transcript is not None:
                yield transcript

    async def close(self) -> None:
        if self._ws is not None:
            try:
                await self._ws.send(json.dumps({"type": "CloseStream"}))
            except Exception:
                pass
            await self._ws.close()
            self._ws = None
